splice_encoding/src/main/java/com/splicemachine/utils/logging/MaskPatternLayout.java

```diff
@@ -0,0 +1,67 @@
+package com.splicemachine.utils.logging;
+
+import org.apache.log4j.Logger;
+import org.apache.log4j.PatternLayout;
+import org.apache.log4j.spi.LoggingEvent;
+
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public class MaskPatternLayout extends PatternLayout {
+
+    private static final String MASK = "***MASKED SENSITIVE INFO***";
+    private Pattern maskPattern = null;
+
+    public String getMaskPattern() {
+        return maskPattern.toString();
+    }
+
+    public void setMaskPattern(String pattern) {
+        maskPattern = Pattern.compile(pattern);
+    }
+
+    @Override
+    public String format(LoggingEvent event) {
+        if (maskPattern == null || !(event.getMessage() instanceof String)) {
+            return super.format(event);
+        }
+
+        String message = event.getRenderedMessage();
+        Matcher matcher = maskPattern.matcher(message);
+
+        if (matcher.find()) {
+            String maskedMessage = "";
+            for(int i = 0; i <= matcher.groupCount(); i++) {
+                int start;
+                int end;
+                if (i == 0) {
+                    start = 0;
+                } else {
+                    start = matcher.end(i);
+                    maskedMessage += MASK;
+                }
+                if (i == matcher.groupCount()) {
+                    end = message.length()-1;
+                } else {
+                    end = matcher.start(i+1) - 1;
+                }
+                if (start > end || end >= message.length()) {
+                    continue;
+                }
+                for (int j = start; j <= end; j++) {
+                    maskedMessage += message.charAt(j);
```

Comment:
Please use StringBuilder for better performance.

Reply:
Also, it's probably quicker to append a substring instead of a character at a time.
String has the replaceAll() method.  Maybe an interface such as this could be used.  It's likely already optimized.

Reply:
@changli6 Fixed with `StringBuilder` since this will not have race condition. @msirek replaceAll may replace the wrong string, for example, a regex `abc ([0-9]+)` only matches the second numbers in `123 abc 123`, `replaceAll` may replace the first one.